core/trino-main/src/main/java/io/trino/sql/ExpressionUtils.java

```diff
@@ -265,6 +276,40 @@ public static Expression filterConjuncts(Metadata metadata, Expression expressio
         };
     }
 
+    /**
+     * Returns whether expression is effectively literal. An effectitvely literal expression is a constant value, or null,
+     * in either {@link Literal} form, or other form returned by {@link LiteralEncoder}.
+     */
+    public static boolean isEffectivelyLiteral(PlannerContext plannerContext, Session session, Expression expression)
+    {
+        if (expression instanceof Literal) {
+            return true;
+        }
+        if (expression instanceof Cast) {
+            return ((Cast) expression).getExpression() instanceof Literal
+                    // a Cast(Literal(...)) can fail, so this requires verification
```

Comment:
Why is it important that the cast doesn't fail?

I guess, we don't want the query to fail in the stats calculator, or in `SimplifyCountOverConstant`, trying to evaluate the constant value. 

However, in `PredicatePushDown`, we don't evaluate the expression, so we could consider the potentially unsafe expression as a constant. And in `LocalExecutionPlanner` we shouldn't care if it fails, but take the opportunity to use the shortcut for trivial projections.

Also, `Literal` is not safe. It is possible to create a `GenericLiteral` with mismatching value, and that also fails to evaluate.

I suggest a different approach on handling failures: we could skip the validation here, and if the "effectively literal" value is to be evaluated during the optimization phase, then we could use the safe `ExpressionInterpreter.optimize` method.

Reply:
> Why is it important that the cast doesn't fail?

the `isEffectivelyLiteral` is a boolean-returning method intended to be used wherever we inspect whether an expression can be considered "a literal". It's not intended to throw for a valid input, as it would make the usage harder.

Note that we don't need to be 'smart' about failing cast case. In a typical case, `ExpressionInterpreter` will fold it to a `fail` call (which isn't recognized as a literal). 

> Also, Literal is not safe. It is possible to create a GenericLiteral with mismatching value, and that also fails to evaluate.

True. However, no optimizer / optimizer rule should do this. Literals coming from parser should be validated in ExpressionAnalyzer.

> I suggest a different approach on handling failures: we could skip the validation here, and if the "effectively literal" value is to be evaluated during the optimization phase, then we could use the safe ExpressionInterpreter.optimize method.

That would work. Note however that my intention was to capture literals, and other things produced by LiteralEncoder. Hence the method name & javadoc and that's also why I choose not to fail. 
Do you feel strongly about this?

Reply:
> the isEffectivelyLiteral is a boolean-returning method intended to be used wherever we inspect whether an expression can be considered "a literal". It's not intended to throw for a valid input, as it would make the usage harder.

My question was: why do we try to filter-out failing casts instead of just reporting them as "effectively literals". I wasn't trying to suggest that `isEffectivelyLiteral` should throw. Sorry for not being clear.

> Note that we don't need to be 'smart' about failing cast case. In a typical case, ExpressionInterpreter will fold it to a fail call (which isn't recognized as a literal).

Yes, but we shouldn't depend on the rule order, especially that `isEffectivelyLiteral` can be reused in the future.

> that my intention was to capture literals, and other things produced by LiteralEncoder.

If we drop the failure check in `isEffectivelyLiteral`, which is my suggestion, that intention is still satisfied. It would then depend on the caller whether they care about errors (and if they do, to use `ExpressionInterpreter.optimize`).
Generally, I think that using `ExpressionInterpreter.optimize` instead of `ExpressionInterpreter.evaluate` should be the rule anywhere before the execution.

If we drop the failure check in `isEffectivelyLiteral`, we are also consistent wrt failing casts and failing `GenericLiterals` (until we fix them).

On the other hand, if `isEffectivelyLiteral` reports failing cast as a non-literal, we miss out in PPD and `LocalExecutionPlanner`, as I mentioned before.

Reply:
Considering `LocalExecutionPlanner`, are there tests for the case with `Cast(Literal))` being handled as trivial projection? It was not possible before.

Reply:
As discussed offline, the intended contract of this method is to filter simple constant expressions which do not fail, so that the caller does not need to deal with potential failure.

I consider this a fair decision, as the majority of failing expressions should never reach this point (supposed that they go through the `ExpressionInterpreter` first). So, we don't lose much optimization opportunities by filtering out failing expressions, while the usage of the method is considerably simplified.

However, for the contract to hold, we need to validate Literals also, not only the casts. Maybe add the validation for Literals for now, and a TODO that user-provided Literals should ba validated in the `ExpressionAnalyzer`?

Reply:
> However, for the contract to hold, we need to validate Literals also, not only the casts.

Done in https://github.com/trinodb/trino/pull/10720, so let me skip adding it here.